controllers/notebook/notebook_controller.go

```diff
@@ -0,0 +1,408 @@
+			notebook.Status.Condition = notebookv1alpha1.NotebookTerminated
+			notebook.Status.LastTransitionTime = metav1.Now()
+			notebook.Status.Message = "notebook pod " + pod.Name + " terminated: " + string(pod.Status.Phase)
+			err = r.Status().Update(context.Background(), notebook)
+			if err != nil {
+				r.Log.Error(err, "failed to update notebook status", "notebook", notebook.Name)
+				return ctrl.Result{}, err
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// 2. create the notebook service if not existing
+	nbServiceName := types.NamespacedName{Namespace: notebook.Namespace, Name: nbNamePrefix(notebook.Name)}
+	nbService := &corev1.Service{}
+	err = r.Get(context.Background(), nbServiceName, nbService)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("create notebook service " + nbServiceName.String())
+			nbService = generateService(pod, notebook)
+
+			// create the notebook service
+			err = r.Create(context.Background(), nbService)
+			if err != nil {
+				r.Log.Error(err, "failed to create service "+nbServiceName.String())
+				return reconcile.Result{Requeue: true}, err
+			}
+			r.recorder.Eventf(notebook, core.EventTypeNormal, "SuccessfulCreatedService", "Created service %v", nbService.Name)
+		} else {
+			// for other errors
+			return reconcile.Result{}, err
+		}
+	}
+
+	// 3. create the ingress if not existing
+	nbIngressName := types.NamespacedName{Namespace: notebook.Namespace, Name: nbNamePrefix(notebook.Name)}
+	nbIngress := &v1.Ingress{}
+	err = r.Get(context.Background(), nbIngressName, nbIngress)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("create notebook ingress " + nbIngressName.String())
+			ingress := generateIngress(nbService, notebook, pod)
+
+			// create the notebook service
+			err = r.Create(context.Background(), ingress)
+			if err != nil {
+				r.Log.Error(err, "failed to create ingress "+nbIngressName.String())
+				return reconcile.Result{Requeue: true}, err
+			}
+
+			// update notebook url
+			err = r.Status().Update(context.Background(), notebook)
+		} else {
+			// for other errors
+			return reconcile.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func generateIngress(svc *corev1.Service, nb *notebookv1alpha1.Notebook, pod *corev1.Pod) *v1.Ingress {
+
+	nbPath := path.Join("/", "notebooks", svc.Namespace, svc.Name)
+	prefixPathType := v1.PathTypePrefix
+
+	// http://{host}/notebooks/{namespace}/{svc.Name}/
+	// all traffic requests for this url will be forwarded to the backend service and its endpoint.
+	ingress := &v1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Ingress",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nbNamePrefix(nb.Name),
+			Namespace: nb.Namespace,
+			Labels:    svc.Labels,
+			Annotations: map[string]string{
+				"nginx.ingress.kubernetes.io/rewrite-target": "/",
+				"nginx.org/websocket-services":               fmt.Sprintf(`"%s"`, svc.Name),
```

Comment:
is notebook service uses websocket as its transport protocal ?

Reply:
removed